test/exercises/wavelet/wavelet.chpl

```diff
@@ -1,66 +1,78 @@
-use ecgIO;
-import Math.log2;
+use ecgIO, Math;
 
-config const nLevels = 3;
-config param plotResults = false;
+config const nLevels = 3,
+             plotResults = false;
+
+const SERIALIZE_THRESHOLD = 1024;
 
 proc main() {
     // read sample data
     //  (data pulled from: http://paulbourke.net/dataformats/holter/)
-    const ecgSamples = try! readEcgData("ecg_samples.txt"),
+    const ecgSamples = getEcgSamples("ecg_samples.txt"),
           lenP2 = 2**(floor(log2(ecgSamples.size)):int);
 
     // compute 1D DWT
-    const ecg_dwt = haarWavelet1D(ecgSamples[0..<lenP2], nLevels);
+    const ecgDwt = haarWavelet1D(ecgSamples[0..<lenP2], nLevels);
 
     // generate a plot of the signal and DWT
-    if plotResults then plotDwtData(ecgSamples[0..<lenP2], ecg_dwt, nLevels);
+    if plotResults then plotDwtData(ecgSamples[0..<lenP2], ecgDwt, nLevels);
+
+    const goodCoeffs = getGoodCoeffs("coeffs.good");
+    writeln((+ reduce abs(goodCoeffs - ecgDwt)) < 1e-10);
 }
 
-// compute an n-level wavelet transform of 'x'
-proc haarWavelet1D(x: [?d] ?t, n: int) : [d] t
+// compute an n-level haar wavelet transform of 'x'
+proc haarWavelet1D(x: [?d] ?t, n: int): [d] t
     where d.rank == 1 && isNumericType(t)
 {
     assert(2**log2(d.size) == d.size, "array size must be a power of 2");
 
-    var output : [d] t;
+    var output: [d] t;
     hwRec(x, output, d.size, d.size / 2**n);
     return output;
 }
 
 // recursive helper for 'haarWavelet1D'
-proc hwRec(const signal, ref output, fmax: int, fstop: int) {
+proc hwRec(const signal: [], ref output: [], fmax: int, fstop: int) {
     if fmax == fstop {
         // store the final layer of high-pass coefficients
-        output[{0..<fmax}] = signal;
+        output[0..<fmax] = signal;
     } else {
-        cobegin {
-            // compute and store the low-pass coefficients
-            output[{fmax/2..<fmax}] = downSample2(haarLP(signal));
+        if signal.size > SERIALIZE_THRESHOLD {
+            // compute the next two layers in parallel
+            cobegin {
+                // compute and store the low-pass coefficients
+                output[fmax/2..<fmax] = downSample2(haarLP(signal));
 
-            // compute the high-pass coefficients and start the next layer of filtering
+                // compute the high-pass coefficients and start the next layer of filtering
+                hwRec(downSample2(haarHP(signal)), output, fmax/2, fstop);
+            }
+        } else {
+            // if the problem size is too small, do the same computation sequentially instead
+            output[fmax/2..<fmax] = downSample2(haarLP(signal));
```

Comment:
If I diagnosed it right the first time (and I very well may not have), this still seems like it gives one task 1/2 the work, the next 1/4, the next 1/8, etc.  I think of the typical pattern for a recursively bisected problem as being something more like this sketch:

```chapel
if iShouldDoAllTheWork then
  output[start..<end] = ...;
else {
  const halflen = (end - start + 1) / 2;
  cobegin {
    hwRec(output[start..#halflen]);
    hwRec(output[start+halflen..end]);
  }
}
```

which gives each of the first two tasks 1/2 the work, then they give the next four 1/4 the work, and so on.  

The logic behind "iShouldDoAllTheWork" could be a combination of problem size (like the threshold you've added in this PR) and/or a check on the number of tasks that are currently running (e.g., if all the cores are already busy, maybe I should just do all the work myself rather than creating more tasks that will be serialized anyway?).  This could either be rolled into the original conditional or split in the following way:

```chapel
const len = end - start + 1;
if len > threshold then
  // do all the work myself
else {
  serial (here.runningTasks >= here.maxTaskPar) {
    cobegin {
      hwRec(output[start..#halflen]);
      hwRec(output[start+halflen..end]);
    }
  }
}
```

The slight advantage to this second form is that it can dynamically switch between creating and not creating tasks as some tasks finish and cores become available.  Whether or not this is helpful or not probably depends on how well load balanced the cores are and whether or not there are a power of two of them.

One thing I'm unsure of in this rewrite is whether tasks that are blocked waiting for their child tasks complete will count in that 'runningTasks' count or not.  I.e., say I'm the first task that creates two tasks — are there three running, or two?  (where "2" is the answer we'd like for that throttle...).

Hmm, that also makes me wonder whether this problem benefits from recursive bisection (vs., say, just creating maxTaskParTasks with a coforall and having each carve off its piece of the overall space...  Are we winning anything by using the recursion other than complexity?  (it's admittedly been a long time since I've computed or thought about haar wavelets).

Reply:
> If I diagnosed it right the first time (and I very well may not have), this still seems like it gives one task 1/2 the work, the next 1/4, the next 1/8, etc

That's correct. And this is a specific feature of a discrete wavelet transform, where each successive filtration of the input signal down samples it by a factor of 2, resulting in an array with half the size. See this figure and table for a three level filter bank on [wikipedia's page for DWT](https://en.wikipedia.org/wiki/Discrete_wavelet_transform#Cascading_and_filter_banks).

So I agree that it would be more efficient to split the array in half each time; however, with this example, I was trying to show a pretty direct translation from the mathematical definition of a DWT into parallel Chapel (where the most obvious opportunity for parallelism is to do hierarchically adjacent filtrations on separate tasks (even though the arrays of coefficients aren't the same size)). 

> (e.g., if all the cores are already busy, maybe I should just do all the work myself rather than creating more tasks that will be serialized anyway?). This could either be rolled into the original conditional or split in the following way:

I like this approach. I'll update the example to use that instead.

> One thing I'm unsure of in this rewrite is whether tasks that are blocked waiting for their child tasks complete will count in that 'runningTasks' count or not. I.e., say I'm the first task that creates two tasks — are there three running, or two? (where "2" is the answer we'd like for that throttle...).

I believe elliot answered that question [here](https://github.com/chapel-lang/chapel/issues/20219#issuecomment-1218415408) when we were discussing stabilizing the `runningTasks` method.

> Hmm, that also makes me wonder whether this problem benefits from recursive bisection (vs., say, just creating maxTaskParTasks with a coforall and having each carve off its piece of the overall space.

Each layer of filters needs the results of the previous layer (see that same wikipedia figure), so a recursive bisection felt like an intuitive way to implement it to me. I'm sure there is a more efficient algorithm that I'm not aware of though.